### DISCUSSION TITLE
Best practise for injecting scripts / accessing variables on the page's window

### DISCUSSION THREAD

**Discussion by rcoopr** [wxt-dev/wxt#523] [Q&A] [answered]
Hi, I need to access a variable that is defined on the window of the page that my extension is used on. What is the best method for interacting with that?

I thought setting `world: 'MAIN'` in my content script would give me access to it, but that didn't work.

I found some useful information on [stackoverflow](https://stackoverflow.com/questions/9515704/access-variables-and-functions-defined-in-page-context-from-an-extension) but I'm not totally sure how to translate one of those methods into the wxt-way of doing things.

In case its helpful, ultimately my use case is sending messages to the page's window and executing functions there based off the message contents.

**Comment by aklinker1** [comment, marked as answer]
Firefox doesn't support main world content scripts, so I would recommend not using those, and manually adding a `script` element to the page to support both browsers (method 1).

Right now, WXT doesn't provide any tools to help with this. The only thing you need to do is ensure your injected script is defined as an unlisted script in the entrypoints directory.

```ts
// entrypoints/injected.ts
export default defineUnlistedScript(() => {
  window.dispatchEvent(new CustomEvent("MyEvent", { details: window.someGlobal }));
});
```

Then you can load it as described in method 1 from your link (which is supported by all browsers):

```ts
export default defineContentScript({
  matches: [...],
  main() {
    const s = document.createElement('script');
    s.src = browser.runtime.getURL('/injected.js');
    s.onload =() => s.remove();
    (document.head || document.documentElement).appendChild(s);

    window.addEventListener("MyEvent", console.log);
  },
});
```

This *should* give you access to the global you're looking for.

**Reply by rcoopr** [reply]
Thanks, that makes sense. I have tried it out but run into a network error: `GET chrome-extension://invalid/ net::ERR_FAILED`

So it seems that the `runtime.browser.getURL` is failing to form a valid URL for the resource. I looked for related issues but came up short

**Reply by aklinker1** [reply]
No, `getURL` is fine. You need to add the script to `web_accessible_resources`. The stack overflow post mentions this.
